File: Aplicacao/sendToDB.js
const SerialPort = require('serialport');
const Readline = require('@serialport/parser-readline');
const sqlite3 = require('better-sqlite3');

const sqliteConnection = sqlite3('temperaturas.db');

// Criação da tabela no SQLite (se não existir)
sqliteConnection.prepare(`
  CREATE TABLE IF NOT EXISTS temperaturas (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    temperatura_media REAL NOT NULL,
    timestamp TEXT NOT NULL
  )
`).run();

// Configuração da porta serial
const port = new SerialPort('COM3', { baudRate: 9600 }); // Ajuste 'COM3' para a porta correta
const parser = port.pipe(new Readline({ delimiter: '\r\n' }));

let temperaturas = []; // Array para armazenar as leituras

parser.on('data', (data) => {
  if (data.includes("Temperatura =")) {
    const temperatura = parseFloat(data.split('=')[1].trim());
    temperaturas.push(temperatura);

    // Se já tivermos 10 leituras, calcular a média e subir para o banco de dados
    if (temperaturas.length === 10) {
      const temperaturaMedia = temperaturas.reduce((sum, temp) => sum + temp, 0) / temperaturas.length;
      const timestamp = new Date().toISOString();

      // Inserir a média e a data no SQLite
      try {
        sqliteConnection.prepare('INSERT INTO temperaturas (temperatura_media, timestamp) VALUES (?, ?)').run(temperaturaMedia, timestamp);
        console.log(`Média dos últimos 10 valores (${temperaturaMedia.toFixed(2)}°C) registrada no SQLite com timestamp ${timestamp}`);
      } catch (err) {
        console.error('Erro ao inserir no SQLite:', err);
      }

      // Limpar o array para as próximas 10 leituras
      temperaturas = [];
    }
  }
});

// Fechar conexão ao finalizar
process.on('exit', () => {
  sqliteConnection.close();
});

File: Aplicacao/sendToDB.py
import serial
import sqlite3
from datetime import datetime

# Configuração da conexão com o SQLite
sqlite_connection = sqlite3.connect('temperaturas.db')
cursor = sqlite_connection.cursor()

# Criação da tabela no SQLite (se não existir)
cursor.execute('''
    CREATE TABLE IF NOT EXISTS temperaturas (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        temperatura_media REAL NOT NULL,
        timestamp TEXT NOT NULL
    )
''')
sqlite_connection.commit()

# Configuração da porta serial
# Ajuste 'COM3' para a porta correta (no Linux pode ser algo como '/dev/ttyUSB0')
port = serial.Serial('COM3', baudrate=9600, timeout=1)

temperaturas = []  # Lista para armazenar as leituras

try:
    while True:
        data = port.readline().decode('utf-8').strip()
        
        # Verificar se o dado contém a informação de temperatura
        if "Temperatura =" in data:
            try:
                temperatura = float(data.split('=')[1].strip())
                temperaturas.append(temperatura)
                print(temperatura)

                # Se já tivermos 10 leituras, calcular a média e inserir no banco de dados
                if len(temperaturas) == 5:
                    temperatura_media = sum(temperaturas) / len(temperaturas)
                    timestamp = datetime.now().isoformat()

                    # Inserir a média e a data no SQLite
                    cursor.execute('''
                        INSERT INTO temperaturas (temperatura_media, timestamp)
                        VALUES (?, ?)
                    ''', (temperatura_media, timestamp))
                    sqlite_connection.commit()
                    
                    print(f"Média dos últimos 5 valores ({temperatura_media:.2f}°C) registrada no SQLite com timestamp {timestamp}")

                    # Limpar a lista para as próximas 10 leituras
                    temperaturas = []
            except ValueError:
                print("Erro ao converter a leitura para número.")
except KeyboardInterrupt:
    print("Leitura interrompida pelo usuário.")
finally:
    # Fechar a conexão serial e do banco de dados
    port.close()
    sqlite_connection.close()
